feat: add `blas/base/dtrmm`

ShabiShett07 opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Progresses #2039.

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds the package `blas/base/dtrmm`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #2039

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

ShabiShett07 committed:
feat: add blas/base/dtrmm

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: minor clean-up

Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: add comments

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 commented on lib/node_modules/@stdlib/blas/base/dtrmm/lib/index.js:
```diff
@@ -0,0 +1,68 @@
+'use strict';
+
+/**
+* BLAS routine to perform one of the matrix-matrix operations `B = α * op(A) * B` or `B = α * B * op(A)` where `α` is a scalar, `B` is an `M` by `N` matrix, `A` is a unit, or non-unit, upper or lower triangular matrix and `op( A )` is one of `op( A ) = A` or `op( A ) = A**T`.
```
```suggestion
* BLAS level 3 routine to perform one of the matrix-matrix operations `B = α * op(A) * B` or `B = α * B * op(A)` where `α` is a scalar, `B` is an `M` by `N` matrix, `A` is a unit, or non-unit, upper or lower triangular matrix and `op( A )` is one of `op( A ) = A` or `op( A ) = A**T`.
```

ShabiShett07 commented on lib/node_modules/@stdlib/blas/base/dtrmm/package.json:
```diff
@@ -0,0 +1,68 @@
+    "array",
+    "ndarray",
+    "float64",
+    "single",
```
```suggestion
    "double",
```

ShabiShett07 committed:
chore: minor clean-up

Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
chore: minor clean-up

Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 commented:
@kgryte, I have traversed through the PR once again and made some changes, need your review

kgryte commented on lib/node_modules/@stdlib/blas/base/dtrmm/lib/base.js:
```diff
@@ -0,0 +1,360 @@
+		( !isrma && side === 'right' && uplo === 'lower' && transa === 'no-transpose' )
+	) {
+		for ( j = 0; j < N; j++ ) {
+			for ( k = 0; k < M; k++ ) {
+				ob2 = offsetB + ( k * sb1 ) + ( j * sb0 );
+				tmp = alpha * B[ ob2 ];
+				for ( i = 0; i < k; i++ ) {
+					oa2 = offsetA + ( i * sa1 ) + ( k * sa0 );
+					ob = offsetB + ( i * sb1 ) + ( j * sb0 );
+					B[ ob ] += ( tmp * A[ oa2 ] );
+				}
+				if ( nonunit ) {
+					oa2 = offsetA + ( k * sa1 ) + ( k * sa0 );
+					tmp *= A[ oa2 ];
+				}
+				B[ ob2 ] = tmp;
+			}
+		}
```
```suggestion
		for ( j = 0; j < N; j++ ) {
			ib = offsetB + ( j*sb0 );
			for ( k = 0; k < M; k++ ) {
				ob2 = ib + ( k * sb1 );
				tmp = alpha * B[ ob2 ];
				ia = offsetA + ( k*sa0 );
				for ( i = 0; i < k; i++ ) {
					B[ ib+(i*sb1) ] += ( tmp * A[ ia+(i*sa1) ] );
				}
				if ( nonunit ) {
					tmp *= A[ ia+(k*sa1) ];
				}
				B[ ob2 ] = tmp;
			}
		}
```

kgryte commented on lib/node_modules/@stdlib/blas/base/dtrmm/lib/dtrmm.js:
```diff
@@ -0,0 +1,128 @@
+*
+* @param {string} order - storage layout of `A` and `B`
+* @param {string} side - specifies whether `op( A )` appears on the left or right side of `B`
+* @param {string} uplo - specifies whether the upper or lower triangular part of the matrix `A` is supplied
```
```suggestion
* @param {string} uplo - specifies whether the upper or lower triangular part of the matrix `A` should be referenced
```

kgryte commented on lib/node_modules/@stdlib/blas/base/dtrmm/lib/dtrmm.js:
```diff
@@ -0,0 +1,128 @@
+// MAIN //
+
+/**
+* Performs one of the matrix-matrix operations `B = α * op(A) * B` or `B = α * B * op(A)` where `α` is a scalar, `B` is an `M` by `N` matrix, `A` is a unit, or non-unit, upper or lower triangular matrix and `op( A )` is one of `op( A ) = A` or `op( A ) = A**T`.
```
```suggestion
* Performs one of the matrix-matrix operations `B = α * op(A) * B` or `B = α * B * op(A)` where `α` is a scalar, `B` is an `M` by `N` matrix, `A` is a unit, or non-unit, upper or lower triangular matrix and `op( A )` is one of `op( A ) = A` or `op( A ) = A^T`.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/dtrmm/lib/dtrmm.js:
```diff
@@ -0,0 +1,128 @@
+* @param {string} order - storage layout of `A` and `B`
+* @param {string} side - specifies whether `op( A )` appears on the left or right side of `B`
+* @param {string} uplo - specifies whether the upper or lower triangular part of the matrix `A` is supplied
+* @param {string} transa - specifies the form of `op( A )` to be used in matrix multiplication
```
```suggestion
* @param {string} transa - specifies whether `A` should be transposed, conjugate-transposed, or not transposed
```

ShabiShett07 committed:
refactor: reduce number of operations

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
docs: improve documentation description

ShabiShett07 committed:
refactor: reduce loops and arithmetic operations

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 commented:
@kgryte, i have made changes to the documentation and also reduced the arithmetic opereations as you suggested

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into feature/dtrmm

prajjwalbajpai commented:
@kgryte what work is pending on this PR??

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into feature/dtrmm

kgryte commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

prajjwalbajpai committed:
fix: remove lint errors and update `base.js`

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

prajjwalbajpai committed:
bench: refactor to use string interpolation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

prajjwalbajpai commented:
@kgryte I have fixed the lint errors. I think this PR is good to go now.

kgryte committed:
Merge branch 'develop' into feature/dtrmm
